File: pre1_1.rb
#最大の人数Mと分ける人数N
M, N = 6,6

def check(remain, pre) #remain:残った人数 pre:前のテーブルに配置した人数
    #配置する人がいなくなると終了
    return 0 if remain < 0
    return 1 if remain == 0

    cnt = 0
    pre.upto(M) do |i|
        cnt += check(remain - i, i)
    end
    cnt
end

puts check(N, 2)

File: pre1_1.py
m, n = 10, 100

def check(x, t): # 再帰的処理を行なうために関数化。xが残っている人数で、tが前のテーブルの人数

    if x == 0:
        return 1
    elif x < 0:
        return 0

    cnt = 0
    for i in range(t,m+1): # テーブルに座る人数を求める
        cnt += check(x-i, i)
    return cnt

print(check(n,2))
